include/lsst/meas/extensions/shapeHSM/HsmMomentsControl.h

```diff
@@ -1,46 +1,154 @@
+class HsmSourceMomentsControl : public HsmMomentsControl {
 public:
     LSST_CONTROL_FIELD(badMaskPlanes, std::vector<std::string>, "Mask planes used to reject bad pixels.");
 
-    HsmSourceMomentsControl() : ShapeControl("shape.hsm.moments") {}
-
     HsmSourceMomentsControl(HsmSourceMomentsControl const & other) :
-        algorithms::ShapeControl(other), badMaskPlanes(other.badMaskPlanes) {}
+        badMaskPlanes(other.badMaskPlanes) {}
 
-    HsmSourceMomentsControl(std::string const & name) : algorithms::ShapeControl(name) {
+    HsmSourceMomentsControl() {
         badMaskPlanes.push_back("BAD");
         badMaskPlanes.push_back("SAT");
         badMaskPlanes.push_back("INTRP");
     }
-private:
-    virtual PTR(algorithms::AlgorithmControl) _clone() const;
-    virtual PTR(algorithms::Algorithm) _makeAlgorithm(
-        afw::table::Schema & schema, PTR(daf::base::PropertyList) const & metadata
-    ) const;
 };
 
+class HsmPsfMomentsControl : public HsmMomentsControl {
+public:
+    HsmPsfMomentsControl() {}
+
+    HsmPsfMomentsControl(HsmPsfMomentsControl const & other) {}
+};
 
-class HsmPsfMomentsControl : public algorithms::ShapeControl {
+/// Base class to measure HSM adaptive moments
+///
+/// Use this to consolidate common code for HsmSourceMoments and HsmPsfMoments
+class HsmMomentsAlgorithm : public base::SimpleAlgorithm {
 public:
-    HsmPsfMomentsControl() : ShapeControl("shape.hsm.psfMoments") {}
-    HsmPsfMomentsControl(std::string const & name) : algorithms::ShapeControl(name) {}
-    HsmPsfMomentsControl(HsmPsfMomentsControl const & other) : algorithms::ShapeControl(other) {}
+    /// A typedef to the Control object for this algorithm, defined above.
+    ///  The control object contains the configuration parameters for this algorithm.
+    typedef HsmMomentsControl Control;
+
+    /// @brief Initialize with standard field names and customized documentation.
+    HsmMomentsAlgorithm(Control const & ctrl, std::string const & name, afw::table::Schema & schema, char const* doc) :
+    _doc(doc),
+    _centroidResultKey(
+        base::CentroidResultKey::addFields(schema, name, "HSM Centroid", base::NO_UNCERTAINTY)),
+    _momentsKey(
+        base::ShapeResultKey::addFields(schema, name, "HSM moments", base::NO_UNCERTAINTY)
+    ),
+    _centroidExtractor(schema, name)
+    {}
+
+    /// Calculate moments
+    template<typename PixelT>
+    void calculate(
+        afw::table::SourceRecord& source, // Source for recording moments
+        PTR(afw::image::Image<PixelT>) const& afwImage, // Image on which to measure moments
+        PTR(afw::image::Mask<afw::image::MaskPixel>) const& afwMask, // Mask for image
+        afw::geom::Box2I const& bbox,     // Bounding box
+        afw::geom::Point2D const& center, // Starting center for measuring moments
+        afw::image::MaskPixel const badPixelMask, // Bitmask for bad pixels
+        float const width            // PSF width estimate, for starting moments
+        ) const;
```

Comment:
Now that this is a public class, these parameter comments should be upgraded to full Doxygen comments (either use "//<" or move them to "@param" lines in the function Doxygen.